File: coverage/create-coverage.go
package main

import (
	"bytes"
	"fmt"
	"net/http"
	"os"
)

func main() {

	// Geoserver parameters
	const user = "admin"
	const pwd = "geoserver"
	const gsURL = "http://localhost:8080/geoserver/rest"
	const workspace = "test"
	const style = "dem"
	const coverage = "dem-go"

	client := &http.Client{}

	// 1) Create the coverage
	// Will create a file dem-go.geotiff in data/test/dem-go/
	data, err := os.Open("dem.tiff")
	if err != nil {
		panic(err)
	}
	defer data.Close()

	const url = gsURL + "/workspaces/" + workspace + "/coveragestores/" + coverage + "/file.geotiff"

	request, err := http.NewRequest("PUT", url, data)
	if err != nil {
		panic(err)
	}
	request.SetBasicAuth(user, pwd)

	request.Header.Set("Content-Type", "image/tiff")

	response, err := client.Do(request)
	if err != nil {
		panic(err)
	}
	defer response.Body.Close()
	fmt.Println(response.Status)

	// 2) Change the style
	const xml = "<layer><defaultStyle><name>" + style + "</name></defaultStyle></layer>"
	payload := bytes.NewBufferString(xml)

	request, err = http.NewRequest("PUT", gsURL+"/layers/"+coverage, payload)
	if err != nil {
		panic(err)
	}
	request.SetBasicAuth(user, pwd)

	request.Header.Set("Content-Type", "application/xml; charset=UTF-8")

	response, err = client.Do(request)
	if err != nil {
		panic(err)
	}
	defer response.Body.Close()
	fmt.Println(response.Status)

}

File: coverage/create-coverage.py
import requests
from requests.auth import HTTPBasicAuth


# Geoserver parameters
user = 'admin'
pwd = 'geoserver'
gs_url = 'http://localhost:8080/geoserver/rest'
workspace = 'test'
style = 'dem'
coverage = 'dem-python'

headers_tiff = {'Content-Type': 'image/tiff'}
headers_xml = {'Content-Type': 'application/xml; charset=UTF-8'}

fileobj = open('dem.tiff', 'rb')

# 1) Create the coverage
# Will create a file dem-python.geotiff in data/test/dem-python/
r = requests.put('{0}/workspaces/{1}/coveragestores/{2}/file.geotiff'.format(gs_url, workspace, coverage),
                 auth=HTTPBasicAuth(user, pwd),
                 data=fileobj,
                 headers=headers_tiff
                 )
print(r)
print(r.text)
print('\n')

# 2) Change the style
payload = '<layer><defaultStyle><name>{0}</name></defaultStyle></layer>'.format(
    style)

r = requests.put('{0}/layers/{1}'.format(gs_url, coverage),
                 auth=HTTPBasicAuth(user, pwd),
                 data=payload,
                 headers=headers_xml
                 )
print(r)
print(r.text)